src/libcore/intrinsics.rs

```diff
@@ -1524,7 +1525,7 @@ pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
 
     debug_assert!(is_aligned_and_not_null(src), "attempt to copy from unaligned or null pointer");
     debug_assert!(is_aligned_and_not_null(dst), "attempt to copy to unaligned or null pointer");
-    debug_assert!(!overlaps(src, dst, count), "attempt to copy to overlapping memory");
+    debug_assert!(is_nonoverlapping(src, dst, count), "attempt to copy to overlapping memory");
```

Comment:
Can you split this change into its own PR? That should have none of the problems this does.

Reply:
If we talk about splitting this, I this the best way forward is to remove the `write` change; I doubt `swap_nonoverlapping` is so critical.

Then I can submit a separate PR for the `write` assertion.